crates/data_components/src/mssql/mod.rs

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use async_trait::async_trait;
+use connection_manager::SqlServerConnectionPool;
+use convert::{map_column_type_to_arrow_type, map_type_name_to_column_type};
+use execution_plan::SqlServerExecPlan;
+use futures::StreamExt;
+use snafu::{OptionExt, ResultExt, Snafu};
+
+use arrow::datatypes::{DataType, Field, Schema, SchemaRef};
+use datafusion::{
+    catalog::Session,
+    datasource::{TableProvider, TableType},
+    error::DataFusionError,
+    logical_expr::{Operator, TableProviderFilterPushDown},
+    physical_plan::ExecutionPlan,
+    prelude::Expr,
+    sql::TableReference,
+};
+
+use std::{any::Any, sync::Arc};
+pub mod connection_manager;
+mod convert;
+mod execution_plan;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Error executing query: {source}"))]
+    QueryError { source: tiberius::error::Error },
+
+    #[snafu(display("Invalid connection string: {source}"))]
+    InvalidConnectionStringError { source: tiberius::error::Error },
+
+    #[snafu(display("Unable to connect: {source}"))]
+    SqlServerAccessError { source: tiberius::error::Error },
+
+    #[snafu(display("Unable to connect: {source}"))]
+    ConnectionPoolError {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Failed to retrieve table schema"))]
+    SchemaRetrieval,
+
+    #[snafu(display("Unable to retrieve schema: table '{table}' does not exist"))]
+    SchemaRetrievalTableNotFound { table: String },
+
+    #[snafu(display("Unsupported data type: {data_type}"))]
+    UnsupportedType { data_type: String },
+
+    #[snafu(display("Failed to build record batch: {source}"))]
+    FailedToBuildRecordBatch { source: arrow::error::ArrowError },
+
+    #[snafu(display("No builder found for index {index}"))]
+    NoBuilderForIndex { index: usize },
+
+    #[snafu(display("Failed to downcast builder for {mssql_type}"))]
+    FailedToDowncastBuilder { mssql_type: String },
+
+    #[snafu(display("Unable to generate SQL: {source}"))]
+    UnableToGenerateSQL { source: DataFusionError },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub struct SqlServerTableProvider {
+    #[allow(dead_code)]
```

Comment:
```suggestion
    #[expect(dead_code)]
```

Reply:
Thank you Phillip - confirm I'll remove `#[allow(dead_code)]` in next PR